### DISCUSSION TITLE
How best to import and use OpenStreetMaps or Satellite images as textures on topography?

### DISCUSSION THREAD

**Discussion by rjleveque** [pyvista/pyvista#6364] [Q&A]
### Describe the feature you would like to be added.

@banesullivan asked me to post this question as an issue, and I was also talking with @bjlittle about related things.  Thanks to both of you and the other presenters for the great PyVista tutorial at SciPy 2024 yesterday, and for all the work on PyVista and GeoVista.  They are easy to use and should be very valuable for improving our visualization tools for [GeoClaw](https://www.geoclaw.org), and more generally for other [Clawpack](https://www.clawpack.org/) applications, including 3D wave propagation problems solved with adaptive mesh refinement.

The question I asked Bane was how to easily import an OpenStreetMap image of a specified region into PyVista and then properly use it as a texture when viewing topography as a surface in 3D?  A related question is the best way to import satellite imagery for the same purpose?

I'm attaching a plot below that shows a somewhat successful attempt to drape a satellite image that I captured as a screenshot from Google Earth as a texture.  I'm sure there are better ways, but I've played around with this a bit for various sorts of coastal topography where I already had images from past work (where we used matplotlib to plot color maps of tsunami inundation depth, for example, on such images).  It seems to work well in general with PyVista, particularly for fairly gentle topography.

For this issue I decided to try it with a small piece of coast along the Tacoma Narrows, not far from the Convention Center where SciPy is in full swing.  This is steep topography and although it still worked pretty well, there is one neighborhood that is falling off the cliff Dali-style, perhaps from the way texture works in this case, even though features line up well with the expected elevation elsewhere.  This is a landslide-prone area (in 1949 an earthquake triggered a landslide that created a small tsunami) but for now at least that neighborhood should be on top of the bluff.

So I'd be interested to know if there's a better way to (1) download suitable satellite imagery, (2) use it properly as a texture, and (3) do the same with OpenStreetMaps (or other map images) for cases where it's more useful to see a map on the topography than satellite images, or maybe even put streets and labels from a map on top of satellite images?

### Pseudocode or Screenshots

Here's the code I used to make the plot.  The topo DEM and GE image are small enough I'll attach those as well in a zip file, in case someone wants to play around with this.  The TacomaNarrows.asc file is a standard ASCII raster file that could be read with other GIS tools rather than using our `geoclaw.topotools`.

```
from pylab import *
from clawpack.geoclaw import topotools  # from www.clawpack.org
import pyvista as pv

extent = [-122.535, -122.524, 47.29, 47.303]
url = 'https://www.ngdc.noaa.gov/thredds/dodsC/regional/puget_sound_13_navd88_2014.nc'
topofile = 'TacomaNarrows.asc'

try:
    # read the topo file if previously saved:
    topo = topotools.Topography(topofile, topo_type=3)
except:
    # load from NCEI thredds server:
    topo = topotools.read_netcdf(url,extent=extent)
    topo.write(topofile, topo_type=3, header_style='asc', Z_format='%.3f')


# convert x,y to meters, roughly:
x = (topo.x - topo.x[0]) * 111e3 * cos(topo.y.mean()*pi/180)
y = -(topo.y - topo.y[0]) * 111e3
z = array([0.])
X,Y,Z = meshgrid(x, y, z, indexing='ij')

topoxyz = pv.StructuredGrid(X,Y,Z)

B = topo.Z
B = flipud(B) # to agree with texture below
topoxyz.point_data['B'] = B.flatten(order='C')

warpfactor = 5 # vertical exageration 
topowarp = topoxyz.warp_by_scalar('B', factor=warpfactor)

# sea surface:
sea_level = 1.4  # MSL in meters relative to NAVD88
eta = where(B < sea_level, sea_level, nan)
topoxyz.point_data['eta'] = eta.flatten(order='C')

# don't need to warp flat surface, but eventually want to plot tsunami surface:
etawarp = topoxyz.warp_by_scalar('eta', factor=warpfactor)

# GE image as texture:
GE_file ='GE_TacomaNarrows.jpg'
GE_extent = [-122.535, -122.524, 47.29, 47.303]
texture = pv.read_texture(GE_file)

x1,x2 = (asarray(GE_extent[:2]) - topo.x[0]) * 111e3 * cos(topo.y.mean()*pi/180)
y1,y2 = (asarray(GE_extent[2:]) - topo.y[0]) * 111e3

o = (x1, y1, 0.)
u = (x2, y1, 0.)
v = (x1, y2, 0.)

mapped_surf = topowarp.texture_map_to_plane(o, u, v)

# make PyVista plot:

p = pv.Plotter()
p.add_mesh(mapped_surf,texture=texture)  # topo with GE image
p.add_mesh(etawarp,color='c')  # water surface

p.camera_position = 'xz'
p.camera.azimuth = -60
p.camera.elevation = 50
p.show(window_size=(2500,1500))
```

![pyvista_TacomaNarrows](https://github.com/pyvista/pyvista/assets/720122/3a6dfc6a-45b9-4c47-958e-0aec3c3fb3ec)

[TacomaNarrowsData.zip](https://github.com/user-attachments/files/16167856/TacomaNarrowsData.zip)

**Comment by banesullivan** [comment]
Hey @rjleveque! Thought I'd quickly follow up with some tips. Seems like you've mostly got this!

You can use this code to extract a region from a large raster: https://gist.github.com/banesullivan/b076de01faa36059823f109c0c325357

And you can use a WMS XML file to define an image source like open street maps using one of:

- Open street maps: https://github.com/OSGeo/gdal/blob/master/frmts/wms/frmt_wms_openstreetmap_tms.xml
- Google maps: https://github.com/OSGeo/gdal/blob/master/frmts/wms/frmt_wms_googlemaps_tms.xml

I typically need to lower the `TileLevel` field in these files to something like `14` in order to not blow up the size and for these extractions to not take forever.

Once you extract the imagery you desire as a raster from the above workflow, you can follow a process similar to the example you have here or I have on my website for texture mapping large rasters: https://banesullivan.com/pyvista/examples/geological-map.html

Hope this helps! Let us know if you get this to work!

**Reply by rjleveque** [reply]
Thanks @banesullivan!  I'm working on figuring out the XML file approach and will let you know if I have questions.

Update on the texture mapping example I posted that looked strange -- it looks fine with less extreme vertical exaggeration, setting warpfactor=1.  That neighborhood does slope downward towards the water and it should not look like such an extreme cliff.
